### ISSUE TITLE
Installation issue: dbcsr with intel-oneapi-mkl threads=openmp

### ISSUE THREAD

**Issue by RMeli** [LLNL/spack#42368]
### Steps to reproduce the issue

```console
$ spack spec -I dbcsr ^intel-oneapi-mkl threads=openmp
Input spec
--------------------------------
 -   dbcsr
 -       ^intel-oneapi-mkl threads=openmp

Concretized
--------------------------------
 -   dbcsr@2.6.0%gcc@13.2.0~cuda~cuda_arch_35_k20x~ipo+mpi~mpi_f08~opencl~openmp~rocm+shared build_system=cmake build_type=Release generator=ninja smm=libxsmm arch=linux-ubuntu23.10-zen4
[+]      ^cmake@3.27.9%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu23.10-zen4
[+]          ^curl@8.4.0%gcc@13.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu23.10-zen4
[+]              ^nghttp2@1.57.0%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^openssl@3.1.3%gcc@13.2.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu23.10-zen4
[+]                  ^ca-certificates-mozilla@2023-05-30%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^ncurses@6.4%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^zlib-ng@2.1.5%gcc@13.2.0+compat+opt build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]      ^intel-oneapi-mkl@2024.0.0%gcc@13.2.0~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=openmp arch=linux-ubuntu23.10-zen4
[+]          ^intel-tbb@2021.9.0%gcc@13.2.0~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make patches=91755c6 arch=linux-ubuntu23.10-zen4
[+]      ^libxsmm@1.17%gcc@13.2.0~debug~generator~header-only~large_jit_buffer~shared blas=default build_system=makefile arch=linux-ubuntu23.10-zen4
[+]          ^binutils@2.41%gcc@13.2.0+gas~gold~gprofng~headers~interwork+ld~libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs=shared,static arch=linux-ubuntu23.10-zen4
[+]              ^diffutils@3.9%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^libiconv@1.17%gcc@13.2.0 build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]              ^zstd@1.5.5%gcc@13.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu23.10-zen4
[+]          ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^python@3.11.6%gcc@13.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu23.10-zen4
[+]              ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-ubuntu23.10-zen4
[+]              ^expat@2.5.0%gcc@13.2.0+libbsd build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^libbsd@0.11.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                      ^libmd@1.0.4%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^gettext@0.22.4%gcc@13.2.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^tar@1.34%gcc@13.2.0 build_system=autotools zip=pigz arch=linux-ubuntu23.10-zen4
[+]                      ^pigz@2.7%gcc@13.2.0 build_system=makefile arch=linux-ubuntu23.10-zen4
[+]              ^libffi@3.4.4%gcc@13.2.0 build_system=autotools patches=070b1f3 arch=linux-ubuntu23.10-zen4
[+]              ^libxcrypt@4.4.35%gcc@13.2.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu23.10-zen4
[+]              ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu23.10-zen4
[+]              ^sqlite@3.43.2%gcc@13.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^util-linux-uuid@2.38.1%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^xz@5.4.1%gcc@13.2.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]      ^ninja@1.11.1%gcc@13.2.0+re2c build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^re2c@2.2%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]      ^openmpi@4.1.6%gcc@13.2.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~internal-libevent~internal-pmix~java~legacylaunchers~lustre~memchecker~openshmem~orterunprefix+romio+rsh~singularity+static+vt+wrapper-rpath build_system=autotools fabrics=none schedulers=none arch=linux-ubuntu23.10-zen4
[+]          ^hwloc@2.9.1%gcc@13.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]              ^libpciaccess@0.17%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^util-macros@1.19.3%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^libxml2@2.10.3%gcc@13.2.0+pic~python+shared build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^libevent@2.1.12%gcc@13.2.0+openssl build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^autoconf@2.72%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^automake@1.16.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^libtool@2.4.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^numactl@2.0.14%gcc@13.2.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu23.10-zen4
[+]              ^m4@1.4.19%gcc@13.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu23.10-zen4
[+]                  ^libsigsegv@2.14%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^openssh@9.5p1%gcc@13.2.0+gssapi build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^krb5@1.20.1%gcc@13.2.0+shared build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^bison@3.8.2%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^findutils@4.9.0%gcc@13.2.0 build_system=autotools patches=440b954 arch=linux-ubuntu23.10-zen4
[+]              ^libedit@3.1-20210216%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu23.10-zen4
[+]              ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu23.10-zen4
[+]          ^pmix@5.0.1%gcc@13.2.0~docs+pmi_backwards_compatibility~python~restful build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^pkgconf@1.9.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^py-fypp@3.1%gcc@13.2.0 build_system=python_pip arch=linux-ubuntu23.10-zen4
[+]          ^py-pip@23.1.2%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^py-setuptools@68.0.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^py-wheel@0.41.2%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen4

```

### Error message

<details><summary>Error message</summary>

<pre>

</pre></details>

### Information on your system

I get the same error on two different systems:

```
* **Spack:** 0.22.0.dev0 (b70cb60b6581ee4b02f786de516aa6a24457c063)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu23.10-zen4
* **Concretizer:** clingo
```

```
* **Spack:** 0.22.0.dev0 (b988e8a23f9eed7904a7272e51b9d585421a87c9)
* **Python:** 3.6.15
* **Platform:** cray-sles15-broadwell
* **Concretizer:** clingo
```

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/14098085/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/14098086/spack-build-out.txt)

@mtaillefumier. Tagging @rscohn2 since the issue seems to be related with `intel-oneapi-mkl`(with the `threads=openmp` variant). The issue is related to #42310 and #42327.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by RMeli** [comment]
A workaround similar to #42327 is the following
```diff
diff --git a/var/spack/repos/builtin/packages/dbcsr/package.py b/var/spack/repos/builtin/packages/dbcsr/package.py
index f8561aabc1..79aed38be3 100644
--- a/var/spack/repos/builtin/packages/dbcsr/package.py
+++ b/var/spack/repos/builtin/packages/dbcsr/package.py
@@ -133,6 +133,9 @@ def cmake_args(self):
             self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
         ]
 
+        if self.spec.satisfies("^intel-oneapi-mkl threads=openmp"):
+            args += [f"-DCMAKE_SHARED_LINKER_FLAGS={self.compiler.openmp_flag}"]
+
         # Switch necessary as a result of a bug.
         if "@2.1:2.2" in spec:
             args += ["-DBUILD_TESTING=ON"]
```

However, I think this might need to be fixed at a more fundamental level.

**Comment by rscohn2** [comment]
This package queries the libs from blas and lapack. It does not get ld_flags so there is no way for the mkl package to pass back the openmp flag to the dbcsr package.

**Comment by RMeli** [comment]
This seems to have been fixed by #42653, thanks @rscohn2.
